File: listings/thread_safety/src/thread_safety.cpp
#include <cstddef>
#include <cstdlib>
#include <cstdio>
#include <thread>
#include <vector>

constexpr size_t NELEMENTS = 1'000'000;
constexpr size_t NTHREADS = 8;
constexpr size_t QOT = NELEMENTS / NTHREADS;
constexpr size_t REM = NELEMENTS % NTHREADS;

auto main() -> int {
    std::vector<int> vector(NELEMENTS, 1);
    std::vector<std::thread> threads(NTHREADS);

    int result = 0;
    for (size_t t = 0; t < NTHREADS; ++t) {
        size_t const start = t * QOT;
        size_t const end = t == NTHREADS - 1 ? start + QOT + REM : start + QOT;
        threads[t] = std::thread([&]() {
            for (size_t i = start; i < end; ++i) {
                result += vector[i];
            }
        });
    }

    for (auto& t: threads) {
        t.join();
    }

    printf("RESULT: %d\n", result);
    return 0;
}

File: listings/thread_safety/src/thread_safety.rs
const NELEMENTS: usize = 1_000_000;
const NTHREADS: usize = 8;
const QOT: usize = NELEMENTS / NTHREADS;
const REM: usize = NELEMENTS % NTHREADS;

fn main() {
    let array = vec![1; NELEMENTS];
    let mut threads = Vec::with_capacity(NTHREADS);

    let mut result = 0;
    for t in 0..NTHREADS {
        let start = t * QOT;
        let end = if t == NTHREADS - 1 {
            start + QOT + REM
        } else {
            start + QOT
        };
        threads.push(std::thread::spawn(|| {
            for i in start..end {
                result += array[i];
            }
        }));
    }

    for t in threads {
        t.join().unwrap();
    }

    println!("RESULT: {result}");
}
